Miscellaneous improvements to the etcd ListAndWatch implementation

giorio94 opened the pull request:
Please review commit by commit, and refer to the individual messages for additional details.

giorio94 committed:
allocator: replace stop chan with context in ListAndWatch

Currently, the ListAndWatch function implemented by allocator backends
takes both a context and a stop channel as parameters, with the context
being then ignored by all implementations. Let's remove this redundancy
and drop the channel in favor of the context, adapting the different
implementations and call sites accordingly.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
allocator: simplify kvstore ListAndWatch implementation

Now that the exit condition is controlled by the context, we can
simplify the current implementation by dropping the select statement
and simply looping over the events channel until it gets closed.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore: fix potential issues when releasing a store

In preparation for the subsequent commit, rework the store
implementation to not use watcher.Stop() anymore, ensuring
that the underlying etcd stream gets correctly terminated.

This also fixes two possible additional problems. First, a
potential race condition due to unsynchronized access to the
`kvstoreWatcher` variable. Second, a potential deadlock due
to Release acquiring a lock and blocking the parallel drain
from the events channel. It is worth mentioning that both
problems are mostly theoretical and extremely unlikely to
happen in production environments, especially because both
Release and Close are never called except during shutdown.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore: drop support of watcher.Stop in favor of context

Now that all remaining users of watcher.Stop have been converted, let's
remove it completely, so that the only way of stopping the watcher is
cancelling the provided context (or closing the etcd client altogether).
This uniforms the behavior, and removes a potential footgun that caused
deadlocks in the past (fixed by 496ba3082158 ("Fix deadlock in identity
watcher") if the caller stopped reading the events channel before it gets
closed. Additionally, it also ensures that the underlying etcd watcher
is properly terminated, rather than potentially leaking it.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore: make ListAndWatch events channel unbuffered

Currently, the buffer size of the events channel returned by
ListAndWatch is configurable. While most callers already make
it unbuffered, a few configure arbitrarily large buffer sizes.
Yet, this is potentially problematic as it can hide deadlocks
that would only occur at sufficiently high scales to fill the
buffer, rather than becoming apparent in smaller environments,
including CI runs; one example is represented by [1].
Additionally, it partially invalidates the value of the
events_queue_seconds metric, given that in most cases it only
accounts for writing into the buffer. Hence, let's just
always make the channel unbuffered to prevent pitfalls.

[1]: cilium/cilium#30797

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
preflight(migrate): prevent race condition when retrieving identities

Make sure that the watcher terminated before returning, to prevent the
possibility that the identities map keeps being modified in parallel
after having already returned it to the caller.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore: make ListAndWatch return a receive-only events channel

Currently, ListAndWatch returns a Watcher struct that wraps the events
channel and a prefix. However, this approach opens for possible misuse
from the consumers, that can potentially modify the prefix (leading to
race conditions), as well as send entries to the events channel and
close it. Let's make the API more robust by simply returning a receive
only channel that can be used by consumers to read the events from.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

squeed reviewed (commented):
Nice cleanup!

marseel commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -574,50 +574,41 @@ func (k *kvstoreBackend) ListIDs(ctx context.Context) (identityIDs []idpool.ID,
 	}
-
-abort:
-	watcher.Stop()
```
Just want to double-check my understanding. `watcher.Stop()` could have been dropped in previous commit as well as far as I can see?
Closing `watcher.Events` can happen only here: https://github.com/cilium/cilium/blob/db6aae76ff0a03ab4bb1615b488c4dfdaca8565a/pkg/kvstore/etcd.go#L775
which already stops watcher.

giorio94 replied:
Yes, `watcher.Stop()` is already a no-op in that case, because we can now only break out of the loop when the events channel is closed, which means that the watcher is already being stopped. I've essentially not removed it in that commit just to keep the diff minimal, leaving the refactoring of the `ListAndWatch` to the subsequent commit.

marseel commented on pkg/kvstore/etcd.go:
```diff
@@ -771,22 +771,14 @@ func (e *etcdClient) watch(ctx context.Context, w *Watcher) {
 	listSignalSent := false
 
+	scopedLog := e.logger.WithField(fieldPrefix, w.Prefix)
+	scopedLog.Debug("Starting watcher...")
```
nit: I would consider doing it as info log together with initial list done and stopping watcher. I believe these parts are critical to understand what is actually happening.

giorio94 replied:
Agreed, the last commit promotes the main log messages to info level, and adds an additional one in case the watcher returns an error.

giorio94 replied:
This is how they look like at the moment:

```
time="2024-12-02T14:27:49.503946673Z" level=info msg="Creating etcd client" ConfigPath=/var/lib/cilium/clustermesh/clustermesh1 KeepAliveHeartbeat=15s KeepAliveTimeout=25s ListLimit=256 MaxInflight=20 RateLimit=20 subsys=kvstore
time="2024-12-02T14:27:49.504749745Z" level=info msg="Connecting to etcd server..." config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" subsys=kvstore
time="2024-12-02T14:27:49.504775179Z" level=info msg="Starting watcher" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/.heartbeat subsys=kvstore
time="2024-12-02T14:28:34.438002343Z" level=info msg="Successfully listed keys before starting watcher" config=/var/lib/cilium/clustermesh/clustermesh1 count=1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/.heartbeat revision=52 subsys=kvstore
time="2024-12-02T14:28:34.438128381Z" level=info msg="Initial etcd connection established" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" subsys=kvstore
time="2024-12-02T14:28:34.438188678Z" level=info msg="Starting to watch prefix" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/.heartbeat revision=53 subsys=kvstore
time="2024-12-02T14:28:34.440929764Z" level=info msg="Starting watcher" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/synced/clustermesh1/ subsys=kvstore
time="2024-12-02T14:28:34.44302509Z" level=info msg="Successfully listed keys before starting watcher" config=/var/lib/cilium/clustermesh/clustermesh1 count=5 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/synced/clustermesh1/ revision=52 subsys=kvstore
time="2024-12-02T14:28:34.443403511Z" level=info msg="Starting watcher" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/identities/v1/clustermesh1/id subsys=kvstore
time="2024-12-02T14:28:34.443793992Z" level=info msg="Starting watcher" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/nodes/v1/clustermesh1/ subsys=kvstore
time="2024-12-02T14:28:34.443349082Z" level=info msg="Starting watcher" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/ip/v1/clustermesh1/ subsys=kvstore
time="2024-12-02T14:28:34.444454817Z" level=info msg="Starting to watch prefix" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/synced/clustermesh1/ revision=53 subsys=kvstore
time="2024-12-02T14:28:34.444622696Z" level=info msg="Starting watcher" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/services/v1/clustermesh1/ subsys=kvstore
time="2024-12-02T14:28:34.446226643Z" level=info msg="Successfully listed keys before starting watcher" config=/var/lib/cilium/clustermesh/clustermesh1 count=3 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/identities/v1/clustermesh1/id revision=52 subsys=kvstore
time="2024-12-02T14:28:34.446592163Z" level=info msg="Starting to watch prefix" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/identities/v1/clustermesh1/id revision=53 subsys=kvstore
time="2024-12-02T14:28:34.451134249Z" level=info msg="Successfully listed keys before starting watcher" config=/var/lib/cilium/clustermesh/clustermesh1 count=2 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/nodes/v1/clustermesh1/ revision=52 subsys=kvstore
time="2024-12-02T14:28:34.452692292Z" level=info msg="Successfully listed keys before starting watcher" config=/var/lib/cilium/clustermesh/clustermesh1 count=0 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/services/v1/clustermesh1/ revision=52 subsys=kvstore
time="2024-12-02T14:28:34.452750877Z" level=info msg="Successfully listed keys before starting watcher" config=/var/lib/cilium/clustermesh/clustermesh1 count=8 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/ip/v1/clustermesh1/ revision=52 subsys=kvstore
time="2024-12-02T14:28:34.452795019Z" level=info msg="Starting to watch prefix" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/services/v1/clustermesh1/ revision=53 subsys=kvstore
time="2024-12-02T14:28:34.456882847Z" level=info msg="Starting to watch prefix" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/nodes/v1/clustermesh1/ revision=53 subsys=kvstore
time="2024-12-02T14:28:34.458586344Z" level=info msg="Starting to watch prefix" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/ip/v1/clustermesh1/ revision=53 subsys=kvstore
time="2024-12-02T14:29:41.289721335Z" level=info msg="Stopped watcher" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/nodes/v1/clustermesh1/ subsys=kvstore
time="2024-12-02T14:29:41.289878746Z" level=info msg="Stopped watcher" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/services/v1/clustermesh1/ subsys=kvstore
time="2024-12-02T14:29:41.289982045Z" level=info msg="Stopped watcher" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/ip/v1/clustermesh1/ subsys=kvstore
time="2024-12-02T14:29:41.290075795Z" level=info msg="Stopped watcher" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/synced/clustermesh1/ subsys=kvstore
time="2024-12-02T14:29:41.290123031Z" level=info msg="Stopped watcher" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/.heartbeat subsys=kvstore
time="2024-12-02T14:29:41.290174885Z" level=info msg="Stopped watcher" config=/var/lib/cilium/clustermesh/clustermesh1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379]" prefix=cilium/cache/identities/v1/clustermesh1/id subsys=kvstore
```

marseel replied:
Very nice, thanks!

marseel commented on pkg/kvstore/events.go:
```diff
@@ -51,31 +51,19 @@ type KeyValueEvent struct {
-	}
-}
+// eventChan is a channel to send event to
+type eventChan chan<- KeyValueEvent
```
nit: I find it quite confusing to use two similar type names for both receive and send events. 
Additionally, there is nothing preventing us from sending event directly to the channel rather than using `emit`.
Let's maybe start with `ReceiveEventChan` and `sendEventChan` (or `emitEventChan` as we shouldn't be sending directly to channel, but calling `emit`). 
Or even better, get rid of types all together and just explicitly use `<-chan KeyValueEvent` or `chan<- KeyValueEvent`

giorio94 replied:
Agreed. I've dropped the send-only `eventChan` and wrapped it into an `emitter` struct, to make more explicit that the internal fields don't need to be directly accessed (as well as drop the `scope` parameter --- `ctx` could be stored as well, but I kept it explicit for the moment, as it looks a bit more idiomatic to me). One extra step could be making `emitter` an interface to completely prevent accessing any internal field, but it seemed overkill to me given that there's a single consumer. WDYT?

marseel reviewed (approved):
Apologies for longer review, it took me multiple attempts to review it :sweat_smile:

giorio94 committed:
kvstore: improve ListAndWatch logging

Improve the log messages of the ListAndWatch function, promoting the
ones corresponding to the main phases of the watch operation to info
level, to simplify troubleshooting possible problems. Additionally,
let's also print a log message in case the etcd watcher returns an
error, before retrying again.

Suggested-by: Marcel Zieba <marcel.zieba@isovalent.com>
Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 commented on pkg/kvstore/etcd.go:
```diff
@@ -907,7 +912,9 @@ reList:
-						scopedLog.WithError(Hint(err)).Debug("Tried watching on compacted revision")
+						scopedLog.WithError(Hint(err)).Info("Tried watching on compacted revision. Triggering relist of all keys")
+					} else {
+						scopedLog.WithError(Hint(err)).Warning("Etcd watcher errored. Triggering relist of all keys")
```
This turned out not being a great idea, as it is also triggered by the clustermesh interceptor when the target clustermesh-apiserver is restarted, in turn getting flagged by CI. Another observed errors is:

```
level=warning msg="Etcd watcher errored. Triggering relist of all keys" config=/var/lib/cilium/clustermesh/cluster1 endpoints="[https://clustermesh-apiserver.kube-system.svc:2379/]" error="rpc error: code = Unknown desc = malformed header: missing HTTP content-type" prefix=cilium/cache/identities/v1/cluster1/id revision=0 subsys=kvstore (1 occurrences)
``` 

I'll try to figure out whether it is possible to ignore these specific cases, or the warning should be demoted/dropped altogether.

giorio94 replied:
Updated that piece of code to only emit a debug log if the error is expected and triggered by an interceptor (I used an extra error type to avoid leaking details of the interceptor into the kvstore implementation). Additionally lowered the severity of generic errors to info level, given that they could potentially happen in clustermesh scenarios when the clustermesh-apiserver is restarted, to avoid issues in CI.

marseel replied:
> Additionally lowered the severity of generic errors to info level, given that they could potentially happen in clustermesh scenarios when the clustermesh-apiserver is restarted, to avoid issues in CI.

I'm not sure about it. If remote service (clustermesh-apiserver) restarts I would definitely expect to have a warning log instead of info log. This is not part of "normal" operations, but we expect it to resolve by itself.

giorio94 replied:
Yeah, I don't terribly like it either. Having it at info level just seemed a decent compromise to me to avoid further increasing the complexity of this PR with trying to figure out all expected cases to demote, considering that it is already an improvement compared to the existing version which doesn't emit any logs in that case.

> If remote service (clustermesh-apiserver) restarts I would definitely expect to have a warning log instead of info log. This is not part of "normal" operations, but we expect it to resolve by itself.

There is actually already a warning log that covers that case (allow-listed in CI), and that is emitted by the clustermesh subsystem when the etcd connection is restarted: https://github.com/cilium/cilium/blob/a3489f190ba6e87b5336ee685fb6c80b1270d06d/pkg/clustermesh/common/remote_cluster.go#L251

gandro reviewed (approved):
This is an awesome cleanup, thank you! No new feedback from my side, all things have already been mentioned.

marseel reviewed (approved):
Thanks!
